nncf/dynamic_graph/context.py

```diff
@@ -235,19 +357,26 @@ def pop_scope(self):
         self.relative_scopes_stack.pop()
         self.scope_modules.pop()
 
-    def register_pre_hooks(self, fn_list: List[Callable], ia_op_exec_context: InputAgnosticOperationExecutionContext):
-        if ia_op_exec_context in self._pre_hooks:
-            raise KeyError("Pre hook for context {} is already registered".format(str(ia_op_exec_context)))
-        self._pre_hooks[ia_op_exec_context] = fn_list
+    def register_pre_hooks(self, fn_list: List[Callable], ia_op_exec_context: InputAgnosticOperationExecutionContext,
+                           input_port_id: int):
+        pre_hook_id = PreHookId(ia_op_exec_context, input_port_id)
+        if pre_hook_id in self._pre_hooks:
+            raise KeyError("Pre hook for context {} is already registered".format(str(pre_hook_id)))
+        self._pre_hooks[pre_hook_id] = fn_list
```

Comment:
1) Pre-hooks is what will allow to quantize branches in the network separately.
Below, on the left, is the result of the MobileNetV2 quantization prior to this PR, and on the right - with this PR.
![Screenshot from 2020-12-01 19-33-00](https://user-images.githubusercontent.com/31695470/100768884-6ee9cc80-340c-11eb-96f4-443a3d341a47.png)

Notice that the `__add__` node used to have a single, nominal pre-hook insertion point, while in fact it takes two tensor activations. The pre-hook insertion points were also non-functional, i.e. the actual insertion could only occur at a post-hook point, which led to a situation where either both of the branches would be quantized with a single quantizer (if a FQ is inserted in the top-most post-hook insertion point), or none of them would. 

With the PR, the node has two pre-hook insertion points, corresponding to the number of inputs (i.e. "ports"). Furthermore, it is possible to separately quantize the `conv2d` input, using the `conv2d`'s pre-hook 0, and the residual `__add__` input, using the `__add__`'s pre-hook 0.